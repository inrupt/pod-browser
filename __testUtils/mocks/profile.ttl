@prefix : <#>.
@prefix solid: <http://www.w3.org/ns/solid/terms#>.
@prefix pro: <./>.
@prefix n0: <http://xmlns.com/foaf/0.1/>.
@prefix schem: <http://schema.org/>.
@prefix n1: <http://www.w3.org/ns/auth/acl#>.
@prefix n2: <https://arquisoft.github.io/>.
@prefix ldp: <http://www.w3.org/ns/ldp#>.
@prefix inbox: </inbox/>.
@prefix sp: <http://www.w3.org/ns/pim/space#>.
@prefix meg: </>.
@prefix c: <https://megoth.solid.community/profile/card#>.
@prefix n: <http://www.w3.org/2006/vcard/ns#>.

pro:card a n0:PersonalProfileDocument; n0:maker :me; n0:primaryTopic :me .

:me
    a schem:Person, n0:Person;
    n0:name "Test Testersen";
    sp:preferencesFile </settings/prefs.ttl>;
    sp:storage meg:;
    solid:account meg:;
    solid:privateTypeIndex </settings/privateTypeIndex.ttl>;
    solid:publicTypeIndex </settings/publicTypeIndex.ttl> .
