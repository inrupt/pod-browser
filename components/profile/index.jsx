/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useState } from "react";
import T from "prop-types";
import { vcard } from "rdf-namespaces";
import Link from "next/link";
import {
  Avatar,
  Box,
  Paper,
  InputLabel,
  createStyles,
} from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import { Container, BackToNavLink } from "@inrupt/prism-react-components";

import { CombinedDataProvider, Text, Image } from "@inrupt/solid-ui-react";

import ContactInfoTable from "./contactInfoTable";
import styles from "./styles";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export default function Profile(props) {
  const { profileIri, editing } = props;
  const [error, setError] = useState(null);

  const classes = useStyles();
  const bem = useBem(classes);

  // TODO replace with toast error or something?
  if (error) {
    return error.toString();
  }

  return (
    <Container>
      <Box mt={2} mb={2}>
        <Link href="/contacts" passHref>
          <BackToNavLink>contacts</BackToNavLink>
        </Link>
      </Box>

      <Paper style={{ maxWidth: "800px" }}>
        <Box p={2}>
          <CombinedDataProvider
            datasetUrl={profileIri}
            thingUrl={profileIri}
            onError={setError}
          >
            <Box alignItems="center" display="flex">
              <Box>
                <Avatar className={classes.avatar}>
                  <Image
                    property={vcard.hasPhoto}
                    width={120}
                    errorComponent={() => (
                      <Avatar
                        className={bem("avatar")}
                        alt="Contact photo placeholder"
                      />
                    )}
                  />
                </Avatar>
              </Box>

              <Box p={2}>
                <h3>
                  <Text property={vcard.fn} />
                </h3>
              </Box>
            </Box>

            <hr />

            <Box mt={2}>
              <Box>
                <InputLabel>Name</InputLabel>
                <Text
                  property={vcard.fn}
                  edit={editing}
                  autosave
                  inputProps={{ className: bem("input") }}
                />
              </Box>

              <Box mt={1}>
                <InputLabel>Role</InputLabel>
                <Text
                  property={vcard.role}
                  edit={editing}
                  inputProps={{ className: bem("input") }}
                  autosave
                />
              </Box>

              <Box mt={1}>
                <InputLabel>Company</InputLabel>
                <Text
                  property={vcard.org}
                  edit={editing}
                  inputProps={{ className: bem("input") }}
                  autosave
                />
              </Box>
            </Box>

            <Box mt={4}>
              <InputLabel>Email Addresses</InputLabel>
              <ContactInfoTable property={vcard.hasEmail} editing={editing} />
            </Box>

            <Box mt={4}>
              <InputLabel>Phone Numbers</InputLabel>
              <ContactInfoTable
                property={vcard.hasTelephone}
                editing={editing}
              />
            </Box>
          </CombinedDataProvider>
        </Box>
      </Paper>
    </Container>
  );
}

Profile.propTypes = {
  profileIri: T.string.isRequired,
  editing: T.bool,
};

Profile.defaultProps = {
  editing: false,
};
