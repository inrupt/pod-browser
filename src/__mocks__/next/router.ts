export default {
  push: jest.fn().mockResolvedValue(null),
};
