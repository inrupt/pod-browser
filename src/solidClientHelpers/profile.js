/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  getSolidDataset,
  getStringNoLocale,
  getThing,
  getUrl,
  asUrl,
  getUrlAll,
} from "@inrupt/solid-client";
import { space, vcard, foaf, rdf, schema } from "rdf-namespaces";

export function displayProfileName({ nickname, name, webId }) {
  if (name) return name;
  if (nickname) return nickname;
  return webId;
}

export function getProfileFromPersonThing(profileThing) {
  return {
    avatar: getUrl(profileThing, vcard.hasPhoto),
    name:
      getStringNoLocale(profileThing, vcard.fn) ||
      getStringNoLocale(profileThing, foaf.name),
    nickname:
      getStringNoLocale(profileThing, vcard.nickname) ||
      getStringNoLocale(profileThing, foaf.nick),
    webId: asUrl(profileThing),
    types: getUrlAll(profileThing, rdf.type),
  };
}

export const TYPE_MAP = {
  [foaf.Person]: getProfileFromPersonThing,
  [schema.Person]: getProfileFromPersonThing,
};

export function getProfileFromThingError(contactThing) {
  return new Error(`Cannot handle profile for contact: ${asUrl(contactThing)}`);
}

export function getProfileFromThing(contactThing) {
  const types = getUrlAll(contactThing, rdf.type);
  const profileFn = TYPE_MAP[types.find((type) => TYPE_MAP[type])];

  if (!profileFn) {
    throw getProfileFromThingError(contactThing);
  }

  return profileFn(contactThing);
}

export function packageProfile(webId, dataset) {
  const profile = getThing(dataset, webId);
  return {
    ...getProfileFromPersonThing(profile),
    webId,
    dataset,
    pods: getUrlAll(profile, space.storage),
  };
}

export async function fetchProfile(webId, fetch) {
  const dataset = await getSolidDataset(webId, { fetch });
  return packageProfile(webId, dataset);
}
