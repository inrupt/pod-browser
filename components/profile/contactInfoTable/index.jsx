/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import T from "prop-types";
import React, { useContext, useState } from "react";
import { rdf, vcard } from "rdf-namespaces";
import clsx from "clsx";
import { makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import {
  addUrl,
  asUrl,
  createThing,
  getSourceUrl,
  getThing,
  getUrlAll,
  removeUrl,
  saveSolidDatasetAt,
  setThing,
  setUrl,
} from "@inrupt/solid-client";

import {
  DatasetContext,
  Table,
  TableColumn,
  useSession,
  useThing,
  Value,
} from "@inrupt/solid-ui-react";

import {
  Box,
  MenuItem,
  Typography,
  FormControl,
  Select,
  createStyles,
} from "@material-ui/core";

import { Table as PrismTable } from "@inrupt/prism-react-components";
import styles from "./styles";
import { chain } from "../../../src/solidClientHelpers/utils";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

const TESTCAFE_ID_DELETE_BUTTON = "profile-address-delete-button";
const TESTCAFE_ID_EXISTING_VALUE = "profile-address-existing-value";
const TESTCAFE_ID_NEW_BUTTON = "profile-address-new-button";
const TESTCAFE_ID_NEW_TYPE = "profile-address-new-type";
const TESTCAFE_ID_NEW_VALUE = "profile-address-new-value";

const CONTACT_TYPE_LABEL_MAP = {
  [vcard.Home]: "Home",
  [vcard.Work]: "Work",
};

export const PREFIX_MAP = {
  [vcard.hasTelephone]: "tel:",
  [vcard.hasEmail]: "mailto:",
};

export const DEFAULT_CONTACT_TYPE = vcard.Home;

export function setupAddContactDetail(
  property,
  newContactType,
  newContactValue,
  dataset,
  profile,
  saveHandler,
  setNewContactValue
) {
  return async () => {
    const prefix = PREFIX_MAP[property];
    const newContactDetail = chain(
      setUrl(createThing(), rdf.type, newContactType),
      (t) => setUrl(t, vcard.value, `${prefix}${newContactValue}`)
    );

    const datasetWithContactDetail = setThing(dataset, newContactDetail);

    const newProfile = addUrl(
      profile,
      property,
      asUrl(newContactDetail, getSourceUrl(dataset))
    );

    await saveHandler(newProfile, datasetWithContactDetail);

    setNewContactValue("");
  };
}

export function setupSaveHandler(fetch, setDataset) {
  return async (newThing, datasetToUpdate) => {
    const savedDataset = await saveSolidDatasetAt(
      getSourceUrl(datasetToUpdate),
      setThing(datasetToUpdate, newThing),
      { fetch }
    );
    setDataset(savedDataset);
  };
}

export function setupRemoveRow(profile, property, saveHandler, dataset) {
  return async (rowThing) => {
    const contactDetailUrl = asUrl(rowThing);
    const newProfile = removeUrl(profile, property, contactDetailUrl);
    await saveHandler(newProfile, dataset);
  };
}

export function setupDeleteButtonCell(editing, removeRow, bem) {
  return () => {
    const { thing: rowThing } = useThing();

    if (!editing) {
      return null;
    }

    return (
      <button
        onClick={() => removeRow(rowThing)}
        className={bem("button", "action")}
        type="button"
        data-testid={TESTCAFE_ID_DELETE_BUTTON}
      >
        Delete
      </button>
    );
  };
}

export default function ContactTable({ editing, property }) {
  const tableClass = PrismTable.useTableClass("table", "inherits");
  const classes = useStyles();
  const bem = useBem(classes);

  const [newContactType, setNewContactType] = useState(DEFAULT_CONTACT_TYPE);
  const [newContactValue, setNewContactValue] = useState("");
  const { fetch } = useSession();
  const { dataset, setDataset } = useContext(DatasetContext);
  const { thing: profile } = useThing();

  const contactDetailUrls = getUrlAll(profile, property);
  const contactDetailThings = contactDetailUrls.map((url) => ({
    dataset,
    thing: getThing(dataset, url),
  }));

  const saveHandler = setupSaveHandler(fetch, setDataset);

  const addContactDetail = setupAddContactDetail(
    property,
    newContactType,
    newContactValue,
    dataset,
    profile,
    saveHandler,
    setNewContactValue
  );

  const removeRow = setupRemoveRow(profile, property, saveHandler, dataset);

  const DeleteButtonCell = setupDeleteButtonCell(editing, removeRow, bem);

  return (
    <>
      <Box mb={1}>
        <Table
          things={contactDetailThings}
          className={clsx(tableClass, bem("table"))}
          getRowProps={() => ({
            className: clsx(bem("table__body-row")),
          })}
        >
          <TableColumn
            property={rdf.type}
            body={({ value }) => {
              const existingTypeValue = CONTACT_TYPE_LABEL_MAP[value] || value;
              return <Typography title={value}>{existingTypeValue}</Typography>;
            }}
            dataType="url"
            header={() => null}
          />

          <TableColumn
            property={vcard.value}
            body={() => (
              <Typography>
                <Value
                  edit={editing}
                  autosave
                  dataType="url"
                  inputProps={{
                    className: bem("input"),
                    "data-testid": TESTCAFE_ID_EXISTING_VALUE,
                  }}
                  property={vcard.value}
                  onSave={(savedDataset) => {
                    setDataset(savedDataset);
                  }}
                />
              </Typography>
            )}
            dataType="url"
            header={() => null}
          />

          <TableColumn
            property={vcard.value}
            body={DeleteButtonCell}
            dataType="url"
            header=""
          />
        </Table>
      </Box>

      {editing && (
        <>
          <Box display="flex" alignItems="center">
            <Box>
              <FormControl className={classes.formControl} variant="outlined">
                <Select
                  value={newContactType}
                  onChange={(e) => setNewContactType(e.target.value)}
                  inputProps={{
                    "data-testid": TESTCAFE_ID_NEW_TYPE,
                  }}
                >
                  {Object.keys(CONTACT_TYPE_LABEL_MAP).map((iri) => (
                    <MenuItem key={iri} value={iri}>
                      {CONTACT_TYPE_LABEL_MAP[iri]}
                    </MenuItem>
                  ))}
                </Select>
              </FormControl>
            </Box>

            <Box width="100%" ml={1} mr={1}>
              <input
                label="Value"
                value={newContactValue}
                onChange={(e) => setNewContactValue(e.target.value)}
                className={bem("input")}
                data-testid={TESTCAFE_ID_NEW_VALUE}
              />
            </Box>

            <Box width="8.25em" textAlign="center">
              <button
                type="button"
                onClick={addContactDetail}
                className={bem("button", "action")}
                data-testid={TESTCAFE_ID_NEW_BUTTON}
              >
                Add
              </button>
            </Box>
          </Box>
        </>
      )}
    </>
  );
}

ContactTable.propTypes = {
  editing: T.bool,
  property: T.string.isRequired,
};

ContactTable.defaultProps = {
  editing: false,
};
