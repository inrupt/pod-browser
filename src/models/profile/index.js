/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { vcard } from "rdf-namespaces";
import {
  addStringNoLocale,
  asUrl,
  getThing,
  getUrl,
} from "@inrupt/solid-client";
import { fetchProfile } from "../../solidClientHelpers/profile";
import { getResource } from "../../solidClientHelpers/resource";
// eslint-disable-next-line import/no-cycle
import { getWebIdUrl } from "../contact/person";

/* Model constants */

/* Model functions */

export async function getProfileForContact(personContactUrl, fetch) {
  const {
    response: { dataset, iri },
  } = await getResource(personContactUrl, fetch);
  const webId = getWebIdUrl(dataset, iri);
  const fetchedProfile = await fetchProfile(webId, fetch);
  return fetchedProfile;
}
export async function getProfilesForPersonContacts(people, fetch) {
  const peopleThings = await Promise.all(
    people.map(({ thing }) => getResource(asUrl(thing), fetch))
  );
  const profileResponses = await Promise.all(
    peopleThings.map(async ({ response: { dataset, iri } }) => {
      const url = getWebIdUrl(dataset, iri);
      return getResource(url, fetch);
    })
  );
  return profileResponses
    .filter(({ error }) => !error)
    .map(({ response }) => response)
    .map(({ dataset, iri }) => {
      const thing = getThing(dataset, iri);
      const avatar = getUrl(thing, vcard.hasPhoto);
      return addStringNoLocale(getThing(thing, iri), vcard.hasPhoto, avatar);
    });
}
