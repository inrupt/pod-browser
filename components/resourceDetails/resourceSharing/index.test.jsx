/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import * as RouterFns from "next/router";
import * as SolidClientFns from "@inrupt/solid-client";
import { mountToJson } from "../../../__testUtils/mountWithTheme";
import * as ProfileHelpers from "../../../src/solidClientHelpers/profile";
import * as PermissionHelpers from "../../../src/solidClientHelpers/permissions";
import SessionContext from "../../../src/contexts/sessionContext";
import ResourceSharing, {
  backToDetailsClick,
  handleAddAgentClick,
  handleAddDefaultPermissions,
  NoThirdPartyPermissions,
  onThirdPartyAccessSubmit,
  ThirdPartyPermissions,
  thirdPartySubmitHandler,
  handleChangeDefaultAgentPermissions,
} from "./index";

describe("ResourceSharing", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  test("it renders the sharing action component", () => {
    const name = "name";
    const iri = "iri";
    const webId = "webId";
    const alias = "Full Control";
    const acl = {
      read: true,
      write: true,
      append: true,
      control: true,
    };
    const permissions = [
      {
        webId,
        alias,
        profile: { webId },
        acl,
      },
      {
        webId: "agentId",
        alias,
        profile: { webId: "agentId" },
        acl,
      },
    ];

    const session = {
      info: {
        webId,
      },
      fetch: jest.fn(),
    };

    jest
      .spyOn(RouterFns, "useRouter")
      .mockReturnValueOnce({ pathname: "/pathname/", replace: jest.fn() });

    jest.spyOn(SolidClientFns, "hasResourceAcl").mockReturnValue(true);

    jest.spyOn(SolidClientFns, "hasAccessibleAcl").mockReturnValue(true);

    jest.spyOn(SolidClientFns, "getResourceAcl").mockReturnValueOnce();

    jest.spyOn(SolidClientFns, "getAgentDefaultAccess").mockReturnValueOnce({
      read: true,
      write: true,
      append: true,
      control: true,
    });

    const tree = mountToJson(
      <SessionContext.Provider value={{ session }}>
        <ResourceSharing
          dataset={{}}
          iri={iri}
          name={name}
          permissions={permissions}
          defaultPermissions={[]}
        />
      </SessionContext.Provider>
    );

    expect(tree).toMatchSnapshot();
  });

  test("it renders the NoThirdPartyPermissions with no 3rd party permissions", () => {
    const name = "name";
    const iri = "iri";
    const webId = "webId";
    const alias = "Full Control";
    const acl = {
      read: true,
      write: true,
      append: true,
      control: true,
    };
    const permissions = [
      {
        webId,
        alias,
        profile: { webId },
        acl,
      },
    ];

    const session = {
      info: {
        webId,
      },
      fetch: jest.fn(),
    };

    jest
      .spyOn(RouterFns, "useRouter")
      .mockReturnValueOnce({ pathname: "/pathname/", replace: jest.fn() });

    jest.spyOn(SolidClientFns, "hasResourceAcl").mockReturnValue(true);

    jest.spyOn(SolidClientFns, "hasAccessibleAcl").mockReturnValue(true);

    jest.spyOn(SolidClientFns, "getResourceAcl").mockReturnValueOnce();

    jest.spyOn(SolidClientFns, "getAgentDefaultAccess").mockReturnValueOnce({
      read: true,
      write: true,
      append: true,
      control: true,
    });

    const tree = mountToJson(
      <SessionContext.Provider value={{ session }}>
        <ResourceSharing
          iri={iri}
          name={name}
          permissions={permissions}
          defaultPermissions={[]}
        />
      </SessionContext.Provider>
    );

    expect(tree).toMatchSnapshot();
  });
});

describe("handleAddAgentClick", () => {
  test("it fetches the profile for the webId and adds it to the addedAgents", async () => {
    const webId = "webId";
    const avatar = "avatar";
    const name = "name";
    const setAddedAgents = jest.fn();
    const fetch = jest.fn();
    const profile = { webId, avatar, name };
    const { ACL } = PermissionHelpers;

    jest.spyOn(ProfileHelpers, "fetchProfile").mockResolvedValueOnce(profile);

    const handler = handleAddAgentClick([], setAddedAgents, fetch);
    await handler(webId);

    expect(setAddedAgents).toHaveBeenCalledWith([
      {
        webId,
        profile,
        alias: ACL.NONE.alias,
        acl: ACL.NONE.acl,
      },
    ]);
  });

  test("it does not add duplicate agents", async () => {
    const webId = "webId";
    const avatar = "avatar";
    const name = "name";
    const fetch = jest.fn();
    const setAddedAgents = jest.fn();
    const profile = { webId, avatar, name };

    jest.spyOn(ProfileHelpers, "fetchProfile").mockResolvedValueOnce(profile);

    const handler = handleAddAgentClick([profile], setAddedAgents, fetch);
    await handler(webId);

    expect(setAddedAgents).not.toHaveBeenCalled();
  });

  test("it logs an error when something goes wrong", async () => {
    const fetch = jest.fn();
    jest.spyOn(console, "error").mockImplementationOnce(jest.fn());
    jest.spyOn(ProfileHelpers, "fetchProfile").mockImplementationOnce(() => {
      throw new Error("boom");
    });

    const handler = handleAddAgentClick([], jest.fn(), fetch);
    await handler("agentId");
    /* eslint-disable no-console */
    expect(console.error).toHaveBeenCalledWith("boom");
  });
});

describe("NoThirdPartyPermissions", () => {
  test("it renders a no 3rd party permissions component", () => {
    const classes = {
      centeredSection: "centeredSection",
      listItem: "listItem",
      detailText: "detailText",
    };

    const tree = mountToJson(<NoThirdPartyPermissions classes={classes} />);

    expect(tree).toMatchSnapshot();
  });
});

describe("ThirdPartyPermissions", () => {
  test("it renders null when thirdPartyPermissions is null", () => {
    const iri = "iri";
    const setThirdPartyPermissions = jest.fn();
    const classes = {
      centeredSection: "centeredSection",
      listItem: "listItem",
    };

    expect(
      ThirdPartyPermissions({
        iri,
        setThirdPartyPermissions,
        thirdPartyPermissions: null,
        classes,
      })
    ).toBeNull();
  });

  test("it renders NoThirdPartyPermissions when the list is empty", () => {
    const iri = "iri";
    const setThirdPartyPermissions = jest.fn();
    const thirdPartyPermissions = [];
    const classes = {
      thirdPartyPermissions,
      centeredSection: "centeredSection",
      listItem: "listItem",
    };

    const tree = mountToJson(
      <ThirdPartyPermissions
        iri={iri}
        setThirdPartyPermissions={setThirdPartyPermissions}
        thirdPartyPermissions={null}
        classes={classes}
      />
    );

    expect(tree).toMatchSnapshot();
  });

  test("it renders a list of Permissions if there are permissions in the list", () => {
    const iri = "iri";
    const webId = "webId";
    const avatar = "avatar";
    const name = "name";
    const permission = {
      alias: "Full Control",
      profile: { webId, avatar, name },
      acl: {
        read: true,
        write: true,
        append: true,
        control: true,
      },
    };
    const setThirdPartyPermissions = jest.fn();
    const thirdPartyPermissions = [permission];
    const classes = {
      thirdPartyPermissions,
      centeredSection: "centeredSection",
      listItem: "listItem",
    };

    const tree = mountToJson(
      <ThirdPartyPermissions
        iri={iri}
        setThirdPartyPermissions={setThirdPartyPermissions}
        thirdPartyPermissions={null}
        classes={classes}
      />
    );

    expect(tree).toMatchSnapshot();
  });
});

describe("backToDetailsClick", () => {
  test("it returns a handle to go back to the details action", async () => {
    const resourceIri = "iri";
    const iri = "/pathname/";
    const replace = jest.fn();

    const router = {
      replace,
      pathname: "/pathname/",
      query: {
        resourceIri,
        iri,
      },
    };

    const handler = backToDetailsClick(router);

    await handler();

    expect(replace).toHaveBeenCalledWith(
      {
        pathname: "/resource/[iri]",
        query: { action: "details", resourceIri },
      },
      {
        pathname: "/resource/%2Fpathname%2F",
        query: { action: "details", resourceIri },
      }
    );
  });
});

describe("onThirdPartyAccessSubmit", () => {
  test("it creates a handles that adds an agent to the permissions list", async () => {
    const addedAgents = [];
    const webId = "webId";
    const setAddedAgents = jest.fn();
    const setThirdPartyPermissions = jest.fn();
    const onSubmit = jest.fn();
    const thirdPartyPermissions = [];
    const profile = { webId };
    const { acl, alias } = PermissionHelpers.ACL.CONTROL;
    const permission = {
      alias,
      acl,
      webId,
      profile,
    };

    const handler = onThirdPartyAccessSubmit({
      addedAgents,
      setAddedAgents,
      setThirdPartyPermissions,
      thirdPartyPermissions,
      onSubmit,
    });

    await handler(profile, acl);

    expect(setThirdPartyPermissions).toHaveBeenCalledWith([permission]);
  });
});

describe("thirdPartySubmitHandler", () => {
  test("it create a handler to remove an agent that has access revoked", () => {
    const { CONTROL, NONE } = PermissionHelpers.ACL;
    const webId = "webId";
    const profile = { webId };
    const permissionWithAccess = {
      alias: CONTROL.alias,
      acl: CONTROL.acl,
      webId,
      profile,
    };
    const thirdPartyPermissions = [permissionWithAccess];
    const setThirdPartyPermissions = jest.fn();
    const handler = thirdPartySubmitHandler({
      thirdPartyPermissions,
      setThirdPartyPermissions,
    });

    handler(profile, NONE.acl);

    expect(setThirdPartyPermissions).toHaveBeenCalledWith([]);
  });

  test("it does nothing if there is any access", () => {
    const { CONTROL, READ } = PermissionHelpers.ACL;
    const webId = "webId";
    const profile = { webId };
    const permissionWithAccess = {
      alias: CONTROL.alias,
      acl: CONTROL.acl,
      webId,
      profile,
    };
    const thirdPartyPermissions = [permissionWithAccess];
    const setThirdPartyPermissions = jest.fn();
    const handler = thirdPartySubmitHandler({
      thirdPartyPermissions,
      setThirdPartyPermissions,
    });

    handler(profile, READ.acl);

    expect(setThirdPartyPermissions).not.toHaveBeenCalledWith();
  });
});

describe("handleAddDefaultPermissions", () => {
  test("it addes the added agent to the defaultAgents", () => {
    const { ACL } = PermissionHelpers;
    const defaultAgents = [];
    const setDefaultAgents = jest.fn();
    const webId = "webId";
    const { alias, acl } = ACL.CONTROL;
    const profile = { webId };
    const handler = handleAddDefaultPermissions({
      defaultAgents,
      setDefaultAgents,
    });
    const permission = {
      profile,
      acl,
      alias,
      webId,
    };

    handler(profile, acl);

    expect(setDefaultAgents).toHaveBeenCalledWith([permission]);
  });
});

describe("handleChangeDefaultAgentPermissions", () => {
  test("it returns a handler that removes agenst with revoked permissions", () => {
    const { acl } = PermissionHelpers.ACL.NONE;
    const webId = "webId";
    const agent = { webId };
    const defaultAgents = [agent];
    const setDefaultAgents = jest.fn();
    const handler = handleChangeDefaultAgentPermissions({
      defaultAgents,
      setDefaultAgents,
    });

    handler(agent, acl);

    expect(setDefaultAgents).toHaveBeenCalledWith([]);
  });
});
