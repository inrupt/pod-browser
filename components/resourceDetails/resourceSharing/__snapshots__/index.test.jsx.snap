// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentAccessList it renders a spinner while loading permissions for access control 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-colorPrimary PodBrowser-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="PodBrowser-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="PodBrowser-circle PodBrowser-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`AgentAccessList it renders three lists of permissions for editors, viewers and blocked 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-colorPrimary PodBrowser-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="PodBrowser-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="PodBrowser-circle PodBrowser-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;
